Update nxos_vlan.py

markcurr opened the pull request:
##### SUMMARY
fixes https://github.com/ansible/ansible/issues/39471

```interfaces = vlan.get('vlanshowplist-ifidx')```

This results in 'interfaces' being a list, however we then try to run:

```for i in interfaces.split(','): ```

Which causes a crash as a list cannot have a split applied.

##### ISSUE TYPE
 - Bugfix Pull Request


##### COMPONENT NAME
nxos_vlan
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.2
  config file = None
  configured module search path = [u'/home/mark/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION

Play:
```yaml
- name: map VLANs to VNIs
  nxos_vlan:
    vlan_id: "{{ item.value.vlan }}"
    mapped_vni: "{{ item.value.vni }}"
    name: "{{ item.key }}"
  with_dict: "{{ vni_segments }}"
```
Dictionary:
```
vni_segments:
  test1:
    vlan: 50
    vni: 5000
```
<!--- BEFORE CHANGE -->
```
TASK [create VLAN and map to VNI] *******************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'list' object has no attribute 'split'
failed: [N9K3] (item={'value': {u'vlan': 50, u'vni': 5000}, 'key': u'test1'}) => {"changed": false, "item": {"key": "test1", "value": {"vlan": 50, "vni": 5000}}, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_qoyZSz/ansible_module_nxos_vlan.py\", line 615, in <module>\n    main()\n  File \"/tmp/ansible_qoyZSz/ansible_module_nxos_vlan.py\", line 593, in main\n    have = map_config_to_obj(module, os_platform)\n  File \"/tmp/ansible_qoyZSz/ansible_module_nxos_vlan.py\", line 512, in map_config_to_obj\n    obj = parse_vlan_options(module, os_platform, output, vlan)\n  File \"/tmp/ansible_qoyZSz/ansible_module_nxos_vlan.py\", line 497, in parse_vlan_options\n    obj['interfaces'] = parse_interfaces(module, vlan)\n  File \"/tmp/ansible_qoyZSz/ansible_module_nxos_vlan.py\", line 474, in parse_interfaces\n    for i in interfaces.split(','):\nAttributeError: 'list' object has no attribute 'split'\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
```

AFTER CHANGE
```
TASK [create VLAN and map to VNI] ********************************************************************************************************************************
ok: [N9K3] => (item={'value': {u'vlan': 50, u'vni': 5000}, 'key': u'test1'})
ok: [N9K3] => (item={'value': {u'subnet': u'10.0.0.1/24', u'vlan': 100, u'vni': 10000}, 'key': u'web'})
```

markcurr committed:
Update nxos_vlan.py

trishnaguha commented on lib/ansible/modules/network/nxos/nxos_vlan.py:
```diff
@@ -458,7 +458,7 @@ def parse_interfaces(module, vlan):
     interfaces = vlan.get('vlanshowplist-ifidx')
     if interfaces:
-        for i in interfaces.split(','):
+        for i in interfaces:
```
`interfaces` cannot be a list. The key `'vlanshowplist-ifidx'` returns string only.

trishnaguha commented:
@markcurr Can you share `show vlan | json` output of your device?

markcurr commented:
@trishnaguha This seems to be a symptom of the platform - I'm using N9kV, it looks like it splits the interfaces into a list when there is more than a certain amount. Perhaps the code could be amended to test for the type returned ?

N9K3# sh vlan | json-pretty
{
    "TABLE_vlanbrief": {
        "ROW_vlanbrief": [
            {
                "vlanshowbr-vlanid": "1", 
                "vlanshowbr-vlanid-utf": "1", 
                "vlanshowbr-vlanname": "default", 
                "vlanshowbr-vlanstate": "active", 
                "vlanshowbr-shutstate": "noshutdown", 
                "vlanshowplist-ifidx": [
                    "Ethernet1/3,Ethernet1/4,Ethernet1/6,Ethernet1/7,Ethernet1/8,Ethernet1/9,Ethernet1/10,
Ethernet1/11,Ethernet1/12,Ethernet1/13,Ethernet1/14,Ethernet1/15,Ethernet1/16,Ethernet1/17,Ethernet1/18,Et
hernet1/19,Ethernet1/20,Ethernet1/21,Ethernet1/22,Ethernet1/23,Ethernet1/24,Ethernet1/25,Ethernet1/26,Ethe
rnet1/27,Ethernet1/28,Ethernet1/29,Ethernet1/30,Ethernet1/31,Ethernet1/32,Ethernet1/33,Ethernet1/34,Ethern
et1/35,Ethernet1/36,Ethernet1/37,Ethernet1/38,Ethernet1/39,Ethernet1/40,Ethernet1/41,Ethernet1/42,Ethernet
1/43,Ethernet1/44,Ethernet1/45,Ethernet1/46,Ethernet1/47,Ethernet1/48,Ethernet1/49,Ethernet1/50,Ethernet1/
51,Ethernet1/52,Ethernet1/53,Ethernet1/54,Ethernet1/55,Ethernet1/56,Ethernet1/57,Ethernet1/58,Ethernet1/59
,Ethernet1/60,Ethernet1/61,Ethernet1/62,Ethernet1/63,Ethernet1/64,Ethernet1/65,Ethernet1/66,Ethernet1/67",
 
                    "Ethernet1/68,Ethernet1/69,Ethernet1/70,Ethernet1/71,Ethernet1/72,Ethernet1/73,Etherne
t1/74,Ethernet1/75,Ethernet1/76,Ethernet1/77,Ethernet1/78,Ethernet1/79,Ethernet1/80,Ethernet1/81,Ethernet1
/82,Ethernet1/83,Ethernet1/84,Ethernet1/85,Ethernet1/86,Ethernet1/87,Ethernet1/88,Ethernet1/89,Ethernet1/9
0,Ethernet1/91,Ethernet1/92,Ethernet1/93,Ethernet1/94,Ethernet1/95,Ethernet1/96,Ethernet1/97,Ethernet1/98,
Ethernet1/99,Ethernet1/100,Ethernet1/101,Ethernet1/102,Ethernet1/103,Ethernet1/104,Ethernet1/105,Ethernet1
/106,Ethernet1/107,Ethernet1/108,Ethernet1/109,Ethernet1/110,Ethernet1/111,Ethernet1/112,Ethernet1/113,Eth
ernet1/114,Ethernet1/115,Ethernet1/116,Ethernet1/117,Ethernet1/118,Ethernet1/119,Ethernet1/120,Ethernet1/1
21,Ethernet1/122,Ethernet1/123,Ethernet1/124,Ethernet1/125,Ethernet1/126,Ethernet1/127,Ethernet1/128"
                ]
            }, 
            {
                "vlanshowbr-vlanid": "50", 
                "vlanshowbr-vlanid-utf": "50", 
                "vlanshowbr-vlanname": "test1", 
                "vlanshowbr-vlanstate": "active", 
                "vlanshowbr-shutstate": "noshutdown"
            }, 
            {
                "vlanshowbr-vlanid": "60", 
                "vlanshowbr-vlanid-utf": "60", 
                "vlanshowbr-vlanname": "test2", 
                "vlanshowbr-vlanstate": "active", 
                "vlanshowbr-shutstate": "noshutdown"
            }, 
            {
                "vlanshowbr-vlanid": "100", 
                "vlanshowbr-vlanid-utf": "100", 
                "vlanshowbr-vlanname": "web", 
                "vlanshowbr-vlanstate": "active", 
                "vlanshowbr-shutstate": "noshutdown", 
                "vlanshowplist-ifidx": "Ethernet1/5"
            }, 
            {
                "vlanshowbr-vlanid": "101", 
                "vlanshowbr-vlanid-utf": "101", 
                "vlanshowbr-vlanname": "app", 
                "vlanshowbr-vlanstate": "active", 
                "vlanshowbr-shutstate": "noshutdown"
            }, 
            {
                "vlanshowbr-vlanid": "102", 
                "vlanshowbr-vlanid-utf": "102", 
                "vlanshowbr-vlanname": "db", 
                "vlanshowbr-vlanstate": "active", 
                "vlanshowbr-shutstate": "noshutdown"
            }, 
            {
                "vlanshowbr-vlanid": "103", 
                "vlanshowbr-vlanid-utf": "103", 
                "vlanshowbr-vlanname": "prod", 
                "vlanshowbr-vlanstate": "active", 
                "vlanshowbr-shutstate": "noshutdown"
            }, 
            {
                "vlanshowbr-vlanid": "104", 
                "vlanshowbr-vlanid-utf": "104", 
                "vlanshowbr-vlanname": "octopus", 
                "vlanshowbr-vlanstate": "active", 
                "vlanshowbr-shutstate": "noshutdown"
            }, 
            {
                "vlanshowbr-vlanid": "105", 
                "vlanshowbr-vlanid-utf": "105", 
                "vlanshowbr-vlanname": "bank", 
                "vlanshowbr-vlanstate": "active", 
                "vlanshowbr-shutstate": "noshutdown"
            }
        ]
    }, 
    "TABLE_mtuinfo": {
        "ROW_mtuinfo": [
            {
                "vlanshowinfo-vlanid": "1", 
                "vlanshowinfo-media-type": "enet", 
                "vlanshowinfo-vlanmode": "ce-vlan"
            }, 
            {
                "vlanshowinfo-vlanid": "50", 
                "vlanshowinfo-media-type": "enet", 
                "vlanshowinfo-vlanmode": "ce-vlan"
            }, 
            {
                "vlanshowinfo-vlanid": "60", 
                "vlanshowinfo-media-type": "enet", 
                "vlanshowinfo-vlanmode": "ce-vlan"
            }, 
            {
                "vlanshowinfo-vlanid": "100", 
                "vlanshowinfo-media-type": "enet", 
                "vlanshowinfo-vlanmode": "ce-vlan"
            }, 
            {
                "vlanshowinfo-vlanid": "101", 
                "vlanshowinfo-media-type": "enet", 
                "vlanshowinfo-vlanmode": "ce-vlan"
            }, 
            {
                "vlanshowinfo-vlanid": "102", 
                "vlanshowinfo-media-type": "enet", 
                "vlanshowinfo-vlanmode": "ce-vlan"
            }, 
            {
                "vlanshowinfo-vlanid": "103", 
                "vlanshowinfo-media-type": "enet", 
                "vlanshowinfo-vlanmode": "ce-vlan"
            }, 
            {
                "vlanshowinfo-vlanid": "104", 
                "vlanshowinfo-media-type": "enet", 
                "vlanshowinfo-vlanmode": "ce-vlan"
            }, 
            {
                "vlanshowinfo-vlanid": "105", 
                "vlanshowinfo-media-type": "enet", 
                "vlanshowinfo-vlanmode": "ce-vlan"
            }
        ]
    }
}

trishnaguha commented:
@markcurr looks like list type output is device specific, your change will break the module for other nxos devices.
So I would rather change it this way instead: https://gist.github.com/trishnaguha/c6348b4a1d9079e3b801c15951acf79e

markcurr commented:
Hi Trishna

That looks good to me, Would certainly prevent the issue we see with N9kV 

However (potential) problem I see is only using the first element in the list 
```
if isinstance(interfaces, list):
  vlan_int = get_vlan_int(interfaces[0])
```
We would then miss the other interfaces returned by vlanshowplist-ifidx.

Perhaps something like
```
if isinstance(interfaces, list):
  interfaces_str = ''
  for i in interfaces:
    interfaces_str.join(i)
  vlan_int = get_vlan_int(interfaces[0])
```
What do you think?

trishnaguha commented:
@markcurr what is I see is `vlanshowplist-ifidx` returns all the interfaces as one item as a string to the list.
so the output is like: `v = ["Ethernet1/11,Ethernet1/12,Ethernet1/13,Ethernet1/14"]`
when you do v[0], this returns "Ethernet1/11,Ethernet1/12,Ethernet1/13,Ethernet1/14" which is exactly we want.

But now I see it returns two items after a certain range, so may be something like?
```
+        if isinstance(interfaces, list):
+            interfaces_list = [i.strip() for i in interfaces]
+            interfaces_str = ','.join(interfaces_list)
+            vlan_int = get_vlan_int(interfaces_str)
```

trishnaguha commented:
@markcurr any progress on this?

markcurr commented:
Hi @trishnaguha - had a chance to test, that looks good!

This is what my nxos_vlan.py looks like now

`def get_vlan_int(interfaces):
    vlan_int = []
    for i in interfaces.split(','):
        if 'eth' in i.lower() and '-' in i:
            int_range = i.split('-')
            stop = int((int_range)[1])
            start = int(int_range[0].split('/')[1])
            eth = int_range[0].split('/')[0]
            for r in range(start, stop + 1):
                vlan_int.append(eth + '/' + str(r))
        else:
            vlan_int.append(i)
    return vlan_int


def parse_interfaces(module, vlan):
    vlan_int = []
    interfaces = vlan.get('vlanshowplist-ifidx')
    if interfaces:
        if isinstance(interfaces, list):
            interfaces_list = [i.strip() for i in interfaces]
            interfaces_str = ','.join(interfaces_list)
            vlan_int = get_vlan_int(interfaces_str)

    return vlan_int
`

trishnaguha commented:
@markcurr you are missing the part when interfaces is not of type `list` in parse_interfaces function, please add that before return.

So the overall change should look like this:
```
def get_vlan_int(interfaces):
    vlan_int = []
    for i in interfaces.split(','):
        if 'eth' in i.lower() and '-' in i:
            int_range = i.split('-')
            stop = int((int_range)[1])
            start = int(int_range[0].split('/')[1])
            eth = int_range[0].split('/')[0]
            for r in range(start, stop + 1):
                vlan_int.append(eth + '/' + str(r))
        else:
            vlan_int.append(i)
    return vlan_int


def parse_interfaces(module, vlan):
    vlan_int = []
    interfaces = vlan.get('vlanshowplist-ifidx')
    if interfaces:
        if isinstance(interfaces, list):
            interfaces_list = [i.strip() for i in interfaces]
            interfaces_str = ','.join(interfaces_list)
            vlan_int = get_vlan_int(interfaces_str)
        else:
            vlan_int = get_vlan_int(interfaces)
    return vlan_int
```

trishnaguha commented:
@markcurr Will you be able to push this by Friday as it is module freeze for 2.6?

trishnaguha commented:
Closing this PR as this is not done within time given in roadmap for 2.6. This is addressed in https://github.com/ansible/ansible/pull/40822
